virttest/qemu_devices/qcontainer.py

```diff
@@ -2829,3 +2832,93 @@ def fs_define_by_params(self, name, params, bus=None):
         else:
             raise ValueError("unsupported filesystem driver type")
         return devices
+
+    def tpm_define_by_params(self, name, params):
+        """
+        Create TPM devices by params.
+
+        :param name: Object name of TPM device.
+        :type name: str
+        :param params: TPM device params.
+        :type params: dict
+        :return: List of corresponding TPM devices.
+        :rtype: list
+        """
+        model = params.get("tpm_model")
+        backend = params.get("tpm_backend")
+        version = params.get("tpm_version")
+
+        tmp_dir = data_dir.get_tmp_dir()
+
+        def _handle_log(line):
+            try:
+                utils_misc.log_line('%s_%s_swtpm_setup.log' % (self.vmname, name), line)
+            except Exception as e:
+                logging.warn("Can't log %s_%s_swtpm_setup output: %s.", self.vmname, name, e)
+
+        def _emulator_setup(binary, extra_options=None):
+            setup_cmd = binary
+            if version in ('2.0',):
+                setup_cmd += ' --tpm2'
+
+            tpm_path = os.path.join(tmp_dir, '%s_%s_tpm_state' % (self.vmname, name))
+            if os.path.exists(tpm_path):
+                shutil.rmtree(tpm_path, ignore_errors=True)
```

Comment:
Is it possible that one vm has multiple tpm devices? Will they share this path that has same name?

Reply:
One VM has one tpm device.

Reply:
Okay, thanks @qcheng-redhat